File: templates/problems/09_heap/02_last_stone_weight/solution.go
package last_stone_weight

import "container/heap"

// solveLastStoneWeight simulates the stone smashing game.
// Time: O(n log n), Space: O(n).
func solveLastStoneWeight(stones []int) int {
	h := &solveMaxHeap{}
	for _, s := range stones {
		*h = append(*h, s)
	}
	heap.Init(h)
	for h.Len() > 1 {
		y := heap.Pop(h).(int)
		x := heap.Pop(h).(int)
		if y != x {
			heap.Push(h, y-x)
		}
	}
	if h.Len() == 0 {
		return 0
	}
	return (*h)[0]
}

type solveMaxHeap []int

func (h solveMaxHeap) Len() int            { return len(h) }
func (h solveMaxHeap) Less(i, j int) bool   { return h[i] > h[j] }
func (h solveMaxHeap) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *solveMaxHeap) Push(x interface{})  { *h = append(*h, x.(int)) }
func (h *solveMaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

File: templates/problems/09_heap/02_last_stone_weight/python/solution.py
import heapq


def solve_last_stone_weight(stones: list[int]) -> int:
    """Simulate the stone smashing game.

    Time: O(n log n), Space: O(n).
    """
    # Use a max-heap by negating values.
    heap = [-s for s in stones]
    heapq.heapify(heap)
    while len(heap) > 1:
        y = -heapq.heappop(heap)
        x = -heapq.heappop(heap)
        if y != x:
            heapq.heappush(heap, -(y - x))
    if not heap:
        return 0
    return -heap[0]
